internal/app/app.go

```diff
@@ -1,27 +1,140 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package app
 
 import (
 	"context"
+	"time"
+
+	"github.com/gardener/etcd-wrapper/internal/types"
 
+	"github.com/gardener/etcd-wrapper/internal/bootstrap"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/server/v3/embed"
 	"go.uber.org/zap"
 )
 
 // Application is a top level struct which serves as an entry point for this application.
 type Application struct {
-	ctx        context.Context
-	logger     *zap.Logger
-	caCertPath string
+	ctx              context.Context
+	cancelFn         context.CancelFunc
+	etcdInitializer  bootstrap.EtcdInitializer
+	cfg              *embed.Config
+	etcdClient       *clientv3.Client
+	etcd             *embed.Etcd
+	waitReadyTimeout time.Duration
+	logger           *zap.Logger
+	etcdReady        bool // should have only one actor that updates it, queryAndUpdateEtcdReadiness()
 }
 
-// Setup sets up the application.
+// NewApplication initializes and returns an application struct
+func NewApplication(ctx context.Context, cancelFn context.CancelFunc, brConfig *types.BackupRestoreConfig, waitReadyTimeout time.Duration, logger *zap.Logger) (*Application, error) {
+	etcdInitializer, err := bootstrap.NewEtcdInitializer(brConfig, logger)
+	if err != nil {
+		return nil, err
+	}
+	return &Application{
+		ctx:              ctx,
+		cancelFn:         cancelFn,
+		etcdInitializer:  etcdInitializer,
+		waitReadyTimeout: waitReadyTimeout,
+		logger:           logger,
+	}, nil
+}
+
+// Setup sets up etcd by triggering initialization of the etcd DB.
 func (a *Application) Setup() error {
-	//sets up etcd by calling bootstrap.InitializeEtcd()
+	// Set up etcd
+	cfg, err := a.etcdInitializer.Run(a.ctx)
+	if err != nil {
+		return err
+	}
+	a.cfg = cfg
 	return nil
 }
 
-// Start starts this application.
+// Start sets up readiness probe and starts an embedded etcd.
 func (a *Application) Start() error {
+	var err error
+
+	// Create etcd client for readiness probe
+	cli, err := a.createEtcdClient()
```

Comment:
how can you create a EtcdClient before starting an etcd ? it may work for HA only if you are using service endpoints but afaiu it won't work for non-HA.

Reply:
Why do you say this won't work for non-HA?
This func created a client with `localhost:2379` as the endpoint and ideally should not care if the etcd is up or not unless the client itself is being used to make an API call

Nevertheless, while developing this repo, it has been mainly tested in non-HA mode and has worked perfectly fine.
If you find that this is not working while testing, please report it and this can be reworked

Reply:
ok, may be I'm wrong here, I will re-check whatever I have mentioned